File: src/main/java/com/localtechsupport/cli/model/Technician.java
package com.localtechsupport.cli.model;

import com.fasterxml.jackson.annotation.JsonProperty;
import java.time.LocalDateTime;
import java.util.List;

/**
 * Technician entity from the API
 * 
 * Represents a technician in the tech support system
 */
public class Technician {

    @JsonProperty("id")
    private Long id;

    @JsonProperty("firstName")
    private String firstName;

    @JsonProperty("lastName")
    private String lastName;

    @JsonProperty("email")
    private String email;

    @JsonProperty("phone")
    private String phone;

    @JsonProperty("status")
    private String status; // ACTIVE, ON_VACATION, SICK_LEAVE, TERMINATED

    @JsonProperty("skills")
    private List<String> skills; // List of service types they can handle

    @JsonProperty("fullName")
    private String fullName;

    @JsonProperty("createdAt")
    private LocalDateTime createdAt;

    @JsonProperty("updatedAt")
    private LocalDateTime updatedAt;

    // Additional fields from server
    @JsonProperty("currentWorkload")
    private Integer currentWorkload;

    @JsonProperty("available")
    private Boolean available;

    // Default constructor
    public Technician() {}

    // Getters and Setters
    public Long getId() {
        return id;
    }

    public void setId(Long id) {
        this.id = id;
    }

    public String getFirstName() {
        return firstName;
    }

    public void setFirstName(String firstName) {
        this.firstName = firstName;
    }

    public String getLastName() {
        return lastName;
    }

    public void setLastName(String lastName) {
        this.lastName = lastName;
    }

    public String getEmail() {
        return email;
    }

    public void setEmail(String email) {
        this.email = email;
    }

    public String getPhone() {
        return phone;
    }

    public void setPhone(String phone) {
        this.phone = phone;
    }

    public String getStatus() {
        return status;
    }

    public void setStatus(String status) {
        this.status = status;
    }

    public List<String> getSkills() {
        return skills;
    }

    public void setSkills(List<String> skills) {
        this.skills = skills;
    }

    public String getFullName() {
        return fullName;
    }

    public void setFullName(String fullName) {
        this.fullName = fullName;
    }

    public LocalDateTime getCreatedAt() {
        return createdAt;
    }

    public void setCreatedAt(LocalDateTime createdAt) {
        this.createdAt = createdAt;
    }

    public LocalDateTime getUpdatedAt() {
        return updatedAt;
    }

    public void setUpdatedAt(LocalDateTime updatedAt) {
        this.updatedAt = updatedAt;
    }

    public Integer getCurrentWorkload() {
        return currentWorkload;
    }

    public void setCurrentWorkload(Integer currentWorkload) {
        this.currentWorkload = currentWorkload;
    }

    public Boolean getAvailable() {
        return available;
    }

    public void setAvailable(Boolean available) {
        this.available = available;
    }

    // Constructor with all parameters
    public Technician(Long id, String firstName, String lastName, String email, String phone, 
                      String status, List<String> skills, String fullName, LocalDateTime createdAt, 
                      LocalDateTime updatedAt, Integer currentWorkload, Boolean available) {
        this.id = id;
        this.firstName = firstName;
        this.lastName = lastName;
        this.email = email;
        this.phone = phone;
        this.status = status;
        this.skills = skills;
        this.fullName = fullName;
        this.createdAt = createdAt;
        this.updatedAt = updatedAt;
        this.currentWorkload = currentWorkload;
        this.available = available;
    }

    // Utility methods
    public boolean isActive() {
        return status != null && "ACTIVE".equalsIgnoreCase(status);
    }

    public boolean hasSkill(String serviceType) {
        if (skills == null || serviceType == null || serviceType.trim().isEmpty()) {
            return false;
        }
        String trimmedServiceType = serviceType.trim();
        return skills.stream().anyMatch(skill -> skill != null && skill.trim().equalsIgnoreCase(trimmedServiceType));
    }

    public boolean isAvailableForServiceType(String serviceType) {
        return isActive() && hasSkill(serviceType);
    }

    @Override
    public boolean equals(Object obj) {
        if (this == obj) return true;
        if (obj == null || getClass() != obj.getClass()) return false;
        
        Technician technician = (Technician) obj;
        
        if (id != null ? !id.equals(technician.id) : technician.id != null) return false;
        if (firstName != null ? !firstName.equals(technician.firstName) : technician.firstName != null) return false;
        if (lastName != null ? !lastName.equals(technician.lastName) : technician.lastName != null) return false;
        return email != null ? email.equals(technician.email) : technician.email == null;
    }

    @Override
    public int hashCode() {
        int result = id != null ? id.hashCode() : 0;
        result = 31 * result + (firstName != null ? firstName.hashCode() : 0);
        result = 31 * result + (lastName != null ? lastName.hashCode() : 0);
        result = 31 * result + (email != null ? email.hashCode() : 0);
        return result;
    }

    @Override
    public String toString() {
        return String.format("Technician{id=%d, fullName='%s', email='%s'}", 
                           id, fullName, email);
    }
    
    // Helper methods for formatting and display
    
    /**
     * Get short name for table display (e.g., "John Smith" -> "John S.")
     */
    public String getShortName() {
        if (fullName == null || fullName.trim().isEmpty()) return "Unknown";
        
        String[] parts = fullName.trim().split("\\s+");
        if (parts.length == 1) {
            return parts[0];
        } else if (parts.length >= 2) {
            return parts[0] + " " + parts[1].charAt(0) + ".";
        }
        return fullName;
    }
} 

File: webapp/src/types/Technician.ts
import { TechnicianStatus } from './enums';

/**
 * Technician entity type definition
 * Converted from Java Technician model
 */
export interface Technician {
  id: number;
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  status: TechnicianStatus;
  skills: string[];
  fullName?: string;
  createdAt: string; // ISO date string
  updatedAt: string; // ISO date string
  currentWorkload?: number;
  available?: boolean;
}

/**
 * Technician creation/update request type
 */
export interface TechnicianRequest {
  firstName: string;
  lastName: string;
  email: string;
  phone: string;
  status: TechnicianStatus;
  skills: string[];
}

/**
 * Technician with computed properties for display
 */
export interface TechnicianDisplay extends Technician {
  shortName: string;
  isActive: boolean;
  workloadLevel: 'LOW' | 'MEDIUM' | 'HIGH';
}

/**
 * Utility functions for Technician operations
 */
export const TechnicianUtils = {
  /**
   * Get full name from first and last name
   */
  getFullName: (technician: Pick<Technician, 'firstName' | 'lastName' | 'fullName'>): string => {
    if (technician.fullName?.trim()) {
      return technician.fullName.trim();
    }
    
    const first = technician.firstName?.trim() || '';
    const last = technician.lastName?.trim() || '';
    
    if (!first && !last) return '';
    if (!first) return last;
    if (!last) return first;
    return `${first} ${last}`;
  },

  /**
   * Get short name for table display (e.g., "John Smith" -> "John S.")
   */
  getShortName: (technician: Technician): string => {
    const fullName = TechnicianUtils.getFullName(technician);
    if (!fullName.trim()) return 'Unknown';
    
    const parts = fullName.trim().split(/\s+/);
    if (parts.length === 1) {
      return parts[0];
    } else if (parts.length >= 2) {
      return `${parts[0]} ${parts[1].charAt(0)}.`;
    }
    return fullName;
  },

  /**
   * Check if technician is active
   */
  isActive: (technician: Technician): boolean => {
    return technician.status === TechnicianStatus.ACTIVE;
  },

  /**
   * Check if technician has a specific skill
   */
  hasSkill: (technician: Technician, serviceType: string): boolean => {
    if (!technician.skills || !serviceType?.trim()) {
      return false;
    }
    const trimmedServiceType = serviceType.trim();
    return technician.skills.some(skill => 
      skill?.trim().toLowerCase() === trimmedServiceType.toLowerCase()
    );
  },

  /**
   * Check if technician is available for a service type
   */
  isAvailableForServiceType: (technician: Technician, serviceType: string): boolean => {
    return TechnicianUtils.isActive(technician) && TechnicianUtils.hasSkill(technician, serviceType);
  },

  /**
   * Get workload level based on current workload
   */
  getWorkloadLevel: (technician: Technician): 'LOW' | 'MEDIUM' | 'HIGH' => {
    const workload = technician.currentWorkload || 0;
    if (workload < 3) return 'LOW';
    if (workload < 6) return 'MEDIUM';
    return 'HIGH';
  }
};